src/modules/cron/task.c

```diff
@@ -218,214 +180,223 @@ static void cron_task_rexec_failed (cron_task_t *t, int errnum)
 {
     flux_t *h = t->h;
-    flux_msg_t *msg;
+    flux_future_t *f;
 
-    if (!t->started || t->exited) {
+    if (!t->running || t->exited) {
         errno = EINVAL;
         return -1;
     }
 
-    msg = kill_request_create (t, sig);
-    if (!msg || flux_send (h, msg, 0) < 0) {
-        flux_log_error (h, "cron_task_kill");
+    if (!(f = flux_subprocess_kill (t->p, sig))) {
+        flux_log_error (h, "%s: flux_subprocess_kill", __FUNCTION__);
         return (-1);
     }
+    /* ignore response */
+    flux_future_destroy (f);
     return (0);
 }
 
-static void timeout_cb (flux_reactor_t *r, flux_watcher_t *w,
-                        int revents, void *arg)
-{
-    cron_task_t *t = arg;
-    t->timedout = 1;
-    if (t->timeout_cb)
-        t->timeout_cb (t->h, t, t->arg);
-    else
-        cron_task_kill (t, SIGTERM);
-}
 
-static void cron_task_timeout_start (cron_task_t *t)
-{
-    flux_watcher_t *w;
-    flux_reactor_t *r;
-    if (t->timeout <= 0.0)
-        return;
-    r = flux_get_reactor (t->h);
-    if (!(w = flux_timer_watcher_create (r, t->timeout, 0.0, timeout_cb, t))) {
-        flux_log_error (t->h, "task_timeout_start");
-        return;
-    }
-    flux_watcher_start (w);
-    t->timeout_w = w;
-}
-
-void cron_task_set_timeout (cron_task_t *t, double to, cron_task_state_f cb)
-{
-    t->timeout_cb = cb;
-    t->timeout = to;
-    if (t->started)
-        cron_task_timeout_start (t);
-}
-
-static json_t *exec_request_create (struct cron_task *t,
+static flux_cmd_t *exec_cmd_create (struct cron_task *t,
     const char *command,
     const char *cwd,
     json_t *env)
 {
-    json_t *o;
-    json_t *cmdline = NULL;
+    flux_cmd_t *cmd = NULL;
+    char *tmp_cwd = NULL;
 
-    if ((o = json_object ()) == NULL)
-        return NULL;
-
-    if ((cmdline = json_pack ("[s,s,s]", "sh", "-c", command)) == NULL
-        || (json_object_set_new (o, "cmdline", cmdline) < 0)) {
-        json_decref (cmdline);
-        goto fail;
+    if (!(cmd = flux_cmd_create (0, NULL, NULL))) {
+        flux_log_error (t->h, "exec_cmd_create: flux_cmd_create");
+        goto error;
     }
-
-    if (cwd) {
-        json_t *x = json_string (cwd);
-        if (x == NULL || json_object_set_new (o, "cwd", x) < 0) {
-            json_decref (x);
-            goto fail;
+    if (flux_cmd_argv_append (cmd, "%s", "sh") < 0) {
+        flux_log_error (t->h, "exec_cmd_create: flux_cmd_argv_append");
+        goto error;
+    }
+    if (flux_cmd_argv_append (cmd, "%s", "-c") < 0) {
+        flux_log_error (t->h, "exec_cmd_create: flux_cmd_argv_append");
+        goto error;
+    }
+    if (flux_cmd_argv_append (cmd, "%s", command) < 0) {
+        flux_log_error (t->h, "exec_cmd_create: flux_cmd_argv_append");
+        goto error;
+    }
+    if (!cwd) {
+        /* flux_rexec() requires a cwd */
+        if (!(tmp_cwd = get_current_dir_name ())) {
+            flux_log_error (t->h, "exec_cmd_create: get_get_current_dir_name");
```

Comment:
Maybe gather the broker CWD once at cron module startup instead of every time a task is launched.

Or, consider allowing a default cwd (by inheriting parent cwd) in `flux_rexec`?